k2/python/k2/fsa_algo.py

```diff
@@ -145,3 +164,100 @@ def arc_sort(fsa: Fsa) -> Fsa:
+    a_arc_map = a_arc_map.to(torch.int64) + 1
+    b_arc_map = b_arc_map.to(torch.int64) + 1
+
+    out_fsa = Fsa.from_ragged_arc(ragged_arc)
+    for name, a_value in a_fsa.named_tensor_attr():
+        if name == 'aux_labels':
+            # aux_label is handled specially
+            continue
+        if hasattr(b_fsa, name):
+            b_value = getattr(b_fsa, name)
+
+            # a_arc_map and b_arc_map have been offset by 1
+            # so we need a padding here
+            padding = a_value.new_zeros((1, *a_value.shape[1:]))
+            a_value = torch.cat((padding, a_value), dim=0)
+            b_value = torch.cat((padding, b_value), dim=0)
+
+            assert a_value.dtype == b_value.dtype == torch.float32
+            value = a_value.index_select(0, a_arc_map) \
+                    + b_value.index_select(0, b_arc_map)
+            setattr(out_fsa, name, value)
+
+    if hasattr(b_fsa, 'aux_labels'):
+        b_aux_labels = b_fsa.aux_labels
+        padding = b_aux_labels.new_zeros(1)
+        b_aux_labels = torch.cat((padding, b_aux_labels), dim=0)
+        out_fsa.aux_labels = b_aux_labels.index_select(0, b_arc_map)
+
+    for name, a_value in a_fsa.named_non_tensor_attr():
+        if name == 'osym':
+            continue
+        setattr(out_fsa, name, a_value)
+
+    for name, b_value in b_fsa.named_non_tensor_attr():
+        if not hasattr(out_fsa, name):
+            setattr(out_fsa, name, b_value)
+
+    return out_fsa
+
+
+def shortest_distance(fsa: Fsa) -> torch.Tensor:
+    '''Return the shortest distance from the start state to the final state
+    in the tropical semiring.
+
+    Note:
+      It uses the opposite sign. That is, It uses `max` instead of `min`.
+
+    Args:
+      fsa:
+        The intpu FSA.
+    Returns:
+      The sum of scores along the best path.
+    '''
+    need_arc_map = True
+    _, arc_map = _k2.shortest_distance(fsa.arcs, need_arc_map)
+    arc_map = arc_map.to(torch.int64)
+
+    return fsa.score[arc_map].sum()
+
+
+def shortest_path(fsa: Fsa) -> Fsa:
+    need_arc_map = True
+    ragged_arc, arc_map = _k2.shortest_path(fsa.arcs, need_arc_map)
+    arc_map = arc_map.to(torch.int64)  # required by index_select
+    out_fsa = Fsa.from_ragged_arc(ragged_arc)
+    for name, value in fsa.named_tensor_attr():
+        setattr(out_fsa, name, value.index_select(0, arc_map))
+    for name, value in fsa.named_non_tensor_attr():
+        setattr(out_fsa, name, deepcopy(value))
+    return out_fsa
+
+
+def project_output(fsa: Fsa) -> Fsa:
+    out_fsa = Fsa.from_ragged_arc(fsa.arcs)
+    indexes = out_fsa.arcs.values()[:, 2] != -1
+    out_fsa.arcs.values()[:, 2][indexes] = fsa.aux_labels.to(
+        torch.int32)[indexes]
+    for name, value in fsa.named_tensor_attr():
+        setattr(out_fsa, name, value)
+    for name, value in fsa.named_non_tensor_attr():
+        setattr(out_fsa, name, deepcopy(value))
+    if hasattr(out_fsa, 'osym'):
+        out_fsa.isym = out_fsa.osym
```

Comment:
Ah I see now, 'isym' and 'osym' are the input and output symbol tables.  That was not documented anywhere so should have been commented on!  Again, I don't want to treat these objects as FSTs, let's just call these 'symbol_table' and 'aux_symbol_table' which I think is more self-explanatory.

Reply:
Will rename the symbol tables. They are for visualization only.

Reply:
Regarding symbol tables: one possibility that is more general, is that the symbol table for the symbols on the FSA is stored as a non-tensor attribute 'symbols'; and symbol tables associated with other kinds of labels are stored as 'xxx_symbols' where 'xxx' is the name of the attribute.  Then we could just use names like "aux" or "phones" or whatever we want for the auxiliary symbols, and being int32_t, they'll be treated as symbols.

Reply:
Fixed.

- `symbols` is for `labels`
- `aux_symbols` is for `aux_labels`.